lib/Frontend/DiagnosticVerifier.cpp

```diff
@@ -547,46 +550,103 @@ DiagnosticVerifier::Result DiagnosticVerifier::verifyFile(unsigned BufferID) {
     
     auto &FoundDiagnostic = *FoundDiagnosticIter;
 
-    const char *IncorrectFixit = nullptr;
+    const char *missedFixitLoc = nullptr;
     // Verify that any expected fix-its are present in the diagnostic.
     for (auto fixit : expected.Fixits) {
       // If we found it, we're ok.
-      if (!checkForFixIt(fixit, FoundDiagnostic, InputFile))
-        IncorrectFixit = fixit.StartLoc;
+      if (!checkForFixIt(fixit, FoundDiagnostic, InputFile)) {
+        missedFixitLoc = fixit.StartLoc;
+        break;
+      }
     }
 
-    bool matchedAllFixIts =
-        expected.Fixits.size() == FoundDiagnostic.FixIts.size();
+    bool isUnexpectedFixitsSeen =
+        expected.Fixits.size() < FoundDiagnostic.FixIts.size();
+
+    auto makeActualFixitsPhrase =
+        [&](ArrayRef<DiagnosticInfo::FixIt> actualFixits,
+            std::string &actualFixitsStr) -> std::string {
+      actualFixitsStr = renderFixits(actualFixits, InputFile);
+
+      std::string str("actual fix-it");
+      if (actualFixits.size() >= 2) {
+        str += "s";
+      }
+      str += " seen: ";
+      str += actualFixitsStr;
+      return str;
+    };
+
+    auto emitFixItsError = [&](const char *location, const std::string &message,
+                               const char *replStartLoc, const char *replEndLoc,
+                               const std::string &replStr) {
+      llvm::SMFixIt fix(llvm::SMRange(SMLoc::getFromPointer(replStartLoc),
+                                      SMLoc::getFromPointer(replEndLoc)),
+                        replStr);
+      addError(location, message, fix);
+    };
 
     // If we have any expected fixits that didn't get matched, then they are
     // wrong.  Replace the failed fixit with what actually happened.
-    if (IncorrectFixit) {
+
+    if (missedFixitLoc) {
+      // If we had an incorrect expected fixit, render it and produce a fixit
+      // of our own.
+
+      assert(!expected.Fixits.empty() &&
+             "some fix-its should be expected here");
+
+      const char *replStartLoc = expected.Fixits.front().StartLoc;
+      const char *replEndLoc = expected.Fixits.back().EndLoc;
+
+      std::string message = "expected fix-it not seen";
+      std::string actualFixits;
+
       if (FoundDiagnostic.FixIts.empty()) {
-        addError(IncorrectFixit, "expected fix-it not seen");
+        if (replStartLoc[-1] == ' ') {
+          replStartLoc--;
+        }
```

Comment:
Nit: a comment explaining why this is needed might be nice for future readers.

Reply:
I added comment.
https://github.com/apple/swift/blob/efaacddecf0b309215ff8e41b24d73558c64d6b8/lib/Frontend/DiagnosticVerifier.cpp#L604-L615

I also added here.
https://github.com/apple/swift/blob/efaacddecf0b309215ff8e41b24d73558c64d6b8/lib/Frontend/DiagnosticVerifier.cpp#L650-L661